cmd/evm/internal/t8ntool/block.go

```diff
@@ -0,0 +1,487 @@
+	if err != nil {
+		return NewError(ErrorIO, fmt.Errorf("failed creating output basedir: %v", err))
+	}
+
+	inputData, err := readInput(ctx)
+	if err != nil {
+		return err
+	}
+	block := inputData.ToBlock()
+	block, err = inputData.SealBlock(block)
+	if err != nil {
+		return err
+	}
+
+	return dispatchBlock(ctx, baseDir, block)
+}
+
+func readInput(ctx *cli.Context) (*blockInput, error) {
+	var (
+		headerStr  = ctx.String(InputHeaderFlag.Name)
+		ommersStr  = ctx.String(InputOmmersFlag.Name)
+		txsStr     = ctx.String(InputTxsRlpFlag.Name)
+		cliqueStr  = ctx.String(SealCliqueFlag.Name)
+		ethashOn   = ctx.Bool(SealEthashFlag.Name)
+		ethashDir  = ctx.String(SealEthashDirFlag.Name)
+		ethashMode = ctx.String(SealEthashModeFlag.Name)
+		inputData  = &blockInput{}
+	)
+
+	if ethashOn && cliqueStr != "" {
+		return nil, NewError(ErrorConfig, fmt.Errorf("both ethash and clique sealing specified, only one may be chosen"))
+	}
+
+	if ethashOn {
+		inputData.Ethash = ethashOn
+		inputData.EthashDir = ethashDir
+		switch ethashMode {
+		case "normal":
+			inputData.PowMode = ethash.ModeNormal
+		case "test":
+			inputData.PowMode = ethash.ModeTest
+		default:
+			return nil, NewError(ErrorConfig, fmt.Errorf("unknown pow mode: %s", ethashMode))
+
+		}
+	}
+
+	if headerStr == stdinSelector || ommersStr == stdinSelector || txsStr == stdinSelector || cliqueStr == stdinSelector {
+		decoder := json.NewDecoder(os.Stdin)
+		if err := decoder.Decode(inputData); err != nil {
+			return nil, NewError(ErrorJson, fmt.Errorf("failed unmarshaling stdin: %v", err))
+		}
+	}
+
+	if cliqueStr != stdinSelector && cliqueStr != "" {
+		clique, err := readClique(cliqueStr)
+		if err != nil {
+			return nil, err
+		}
+		inputData.Clique = clique
+	}
+
+	if headerStr != stdinSelector {
+		env, err := readEnv(headerStr)
+		if err != nil {
+			return nil, err
+		}
+		inputData.Env = env
+	}
+
+	if ommersStr != stdinSelector {
+		ommers, err := readOmmers(ommersStr)
+		if err != nil {
+			return nil, err
+		}
+		inputData.OmmersRlp = ommers
+	}
+
+	ommers := []*types.Header{}
+	for _, str := range inputData.OmmersRlp {
+		type extblock struct {
+			Header *types.Header
+			Txs    []*types.Transaction
+			Ommers []*types.Header
+		}
+		var ommer *extblock
+		raw := common.FromHex(str)
+		err := rlp.DecodeBytes(raw, &ommer)
+		if err != nil {
+			return nil, NewError(ErrorRlp, fmt.Errorf("unable to decode ommer from rlp data: %v", err))
+		}
+		ommers = append(ommers, ommer.Header)
+	}
+	inputData.Ommers = ommers
+
+	if txsStr != stdinSelector {
+		txs, err := readTxsRlp(txsStr)
+		if err != nil {
+			return nil, err
+		}
+		inputData.TxRlp = txs
+	}
+
+	txs := []*types.Transaction{}
+	raw := common.FromHex(inputData.TxRlp)
+	err := rlp.DecodeBytes(raw, &txs)
+	if err != nil {
+		return nil, NewError(ErrorRlp, fmt.Errorf("unable to decode transaction from rlp data: %v", err))
+	}
+	inputData.Txs = txs
+
+	return inputData, nil
+}
+
+func readEnv(path string) (*bbEnv, error) {
```

Comment:
These functions are very much repeated. Maybe we can make a utility method that we reuse: 
```golang

func readFile(path, desc string, dest interface{}) error{
       ....
     if err := decoder.Decode(dest); err != nil {
         return nil, NewError(ErrorJson, fmt.Errorf("failed unmarshaling %v from stdin: %v", desc, err))
     ....
}
```
And then just 
```
  env :=  &bbEnv
  err := readFile(path, "env", env)
```

Reply:
Great point, I've updated!